### PR TITLE
[WIP] Add structure for S128 validation, execution

### PR DESCRIPTION
**Pull request by dtig**
* This is dependent on #193 
* Looking for feedback on structuring of operations
* Formalization WIP

### CODE REVIEW

File: document/core/exec/instructions.rst [hunk 1]
```diff
@@ -112,6 +112,12 @@ Where the underlying operators are non-deterministic, because they may return on
    \end{array}
 
 
+Integer, Floating-point Numeric instructions
+............................................
+
+The below subset of numeric instructions is applicable only to operands with :ref:`integer <syntax-int>` and :ref:`floating point <syntax-float>` value types, and not to operands of Fixed-width SIMD value type.
```
Comment (rossberg) [inline document/core/exec/instructions.rst]: This comment and heading seems unnecessary. Either it already follows from syntax, because there are no SIMD testops; or there are, in which case the rules would need explicit side conditions / case distinctions anyway that control what they apply to.

File: document/core/exec/numerics.rst [hunk 2]
```diff
@@ -114,6 +114,10 @@ Floating-Point
 
 where :math:`M = \significand(N)` and :math:`E = \exponent(N)`.
 
+Fixed-width SIMD
+................
+
+:ref:`Fixed-width SIMD values <syntax-simd>` can be used to represent different types of packed data. The format of the this data is :math:`\{t\}\{lwidth\}\times \{lanes\}` and is interpreted as a packed SIMD value of :math:`lanes`number of values of type :math:`t`, which are either :ref:`integers <syntax-int>` or :ref:`floating-point <syntax-float>` numbers of width :math:`lwidth`.
```
Comment (rossberg) [inline document/core/exec/numerics.rst]:
```suggestion
:ref:`Fixed-width SIMD values <syntax-simd>` can be used to represent different types of packed data. The format of this data is :math:`\{t\}\{\X{lwidth}\}\times \{\X{lanes}\}` and is interpreted as a packed SIMD value of :math:`\X{lanes}`number of values of type :math:`t`, which are either :ref:`integers <syntax-int>` or :ref:`floating-point <syntax-float>` numbers of width :math:`\X{lwidth}`.
```
(You need \X to typeset multi-letter var names correctly.)
Also style nit: At most one sentence per line.

However, my real question is that I don't understand the syntax you are suggesting. This ought to define the format for representing SIMD *values*, but the syntax doesn't seem to describe any actual values.

File: document/core/exec/runtime.rst [hunk 3]
```diff
@@ -7,14 +7,14 @@ Runtime Structure
...
-.. index:: ! value, constant, value type, integer, floating-point
+.. index:: ! value, constant, value type, integer, floating-point, fixed-width simd
    pair: abstract syntax; value
 .. _syntax-val:
 
 Values
 ~~~~~~
 
-WebAssembly computations manipulate *values* of the four basic :ref:`value types <syntax-valtype>`: :ref:`integers <syntax-int>` and :ref:`floating-point data <syntax-float>` of 32 or 64 bit width each, respectively.
+WebAssembly computations manipulate *values* of the five basic :ref:`value types <syntax-valtype>`: :ref:`integers <syntax-int>` and :ref:`floating-point data <syntax-float>` of 32 or 64 bit width each  respectively, and :ref: `SIMD data <syntax-simd>` of 128 bit width representing packed integers of floating-point values.
```
Comment (rossberg) [inline document/core/exec/runtime.rst]: "packed integers of floating point values"??? ;)

File: document/core/syntax/values.rst [hunk 4]
```diff
@@ -145,6 +145,19 @@ Conventions
...
+
+Fixed-Width SIMD
+~~~~~~~~~~~~~~~~
+
+Fixed-width SIMD data represents packed integers, or floating-point values. The packed values of a given SIMD value must all be of the same type, which can be any of the signed, unsigned integer types, and floating-point types ocurring in this proposal. The supported representations of the packed values are 4 32-bit, or 2 64-bit |IEEE754|_ numbers, or different widths of packed integer values specifically 2 64-bit integers, 4 32-bit integers, 8 16-bit integers, or 16 8-bit integers.
+
+The number of packed values defined here are lanes :`lanes` can be determined by the bit width :math:`N` of the packed value type.
+
+.. math::
+   lanes = 128/N
```
Comment (rossberg) [inline document/core/syntax/values.rst]:
```suggestion
   \F{lanes}(N) = 128/N
```
(this is a function of N, not a constant)

File: document/core/syntax/values.rst [hunk 5]
```diff
@@ -145,6 +145,19 @@ Conventions
...
 * The meta variable :math:`z` ranges over floating-point values where clear from context.
 
+.. index:: ! simd, fixed-width, vector
+   pair: abstract syntax; packed simd number
+.. _syntax-simd:
+
+Fixed-Width SIMD
+~~~~~~~~~~~~~~~~
+
+Fixed-width SIMD data represents packed integers, or floating-point values. The packed values of a given SIMD value must all be of the same type, which can be any of the signed, unsigned integer types, and floating-point types ocurring in this proposal. The supported representations of the packed values are 4 32-bit, or 2 64-bit |IEEE754|_ numbers, or different widths of packed integer values specifically 2 64-bit integers, 4 32-bit integers, 8 16-bit integers, or 16 8-bit integers.
```
Comment (rossberg) [inline document/core/syntax/values.rst]: Since SIMD values can be interpreted arbitrarily, the abstract syntax needs to pick one unique representation. Similar to integers, this then has to be reinterpreted by explicit meta-operators in contexts that determine a specific interpretation.

File: document/core/valid/instructions.rst [hunk 6]
```diff
@@ -112,6 +112,11 @@ Numeric Instructions
    }
 
 
+Integer, Floating-point Numeric instructions
+............................................
+
+The validation rules for the below subset of numeric instructions are applicable only to operands with :ref:`integer <syntax-int>` and :ref:`floating point <syntax-float>` value types, and not to operands of Fixed-width SIMD value type.
```
Comment (rossberg) [inline document/core/valid/instructions.rst]: Same comment as for numeric.rst applies.

File: document/core/exec/instructions.rst [hunk 7]
```diff
@@ -388,6 +399,93 @@ Memory Instructions
    However, it may be substantially slower on some hardware.
 
 
+.. _exec-memory.size:
```
Comment (rossberg) [inline document/core/exec/instructions.rst]: I would strongly recommend not moving unrelated stuff around for now, since that will cause major headaches with continuous merging of changes from upstream (especially with the memory ops). It also makes reviewing more difficult.

If you want to move stuff for clean-up, maybe insert a `todo::` as a reminder to do that once the proposal is closer to merging.

File: document/core/exec/instructions.rst [hunk 8]
```diff
@@ -388,6 +399,93 @@ Memory Instructions
...
+.. note::
+   The |MEMORYGROW| instruction is non-deterministic.
+   It may either succeed, returning the old memory size :math:`\X{sz}`,
+   or fail, returning :math:`{-1}`.
+   Failure *must* occur if the referenced memory instance has a maximum size defined that would be exceeded.
+   However, failure *can* occur in other cases as well.
+   In practice, the choice depends on the :ref:`resources <impl-exec>` available to the :ref:`embedder <embedder>`.
+
+
+The execution specified for :math:`t\K{.}\LOAD{N}\K{\_}\sx~\memarg` and :math:`t\K{.}\STORE{N}~\memarg` memory instructions is applicable only to operands with :ref:`integer <syntax-int>` and :ref:`floating point <syntax-float>` value types, and not to operands of :ref:`Fixed-width SIMD <syntax-simd>` value type.
```
Comment (rossberg) [inline document/core/exec/instructions.rst]: Same comment as in numerics.rst applies.

File: document/core/exec/runtime.rst [hunk 9]
```diff
@@ -7,14 +7,14 @@ Runtime Structure
 :ref:`Store <store>`, :ref:`stack <stack>`, and other *runtime structure* forming the WebAssembly abstract machine, such as :ref:`values <syntax-val>` or :ref:`module instances <syntax-moduleinst>`, are made precise in terms of additional auxiliary syntax.
 
 
-.. index:: ! value, constant, value type, integer, floating-point
+.. index:: ! value, constant, value type, integer, floating-point, fixed-width simd
```
Comment (rossberg) [inline document/core/exec/runtime.rst]:
```suggestion
.. index:: ! value, constant, value type, integer, floating-point, simd
```

File: document/core/exec/numerics.rst (hunk 2 above)
Comment (dtig) [reply document/core/exec/numerics.rst]: Will be addressing other comments as well, but to address the general issue of syntax which is comes up in other review comments as well is that this PR is dependent on #193 where I've added some notion of a SIMD type, and some basic text for SIMD values, but it's unclear to me where the right place for the different representations is, I'd like to add some expressiveness for the different representations so this can reference that more accurately, and it will be easier to differentiate different operations that are applicable to different types.

**Comment by dtig** [comment]
I let this stagnate, and a better version of this has been merged by @ngzhian, closing.
